src/core/app-template.cc

```diff
@@ -107,11 +108,19 @@ const app_template::seastar_options& app_template::options() const {
 app_template::configuration_reader app_template::get_default_configuration_reader() {
     return [this] (bpo::variables_map& configuration) {
         auto home = std::getenv("HOME");
+
+        std::string config_file = _opts.config_file.get_value();
+        if (config_file[0] == '~' && home) {
```

Comment:
sorry for being paranoid, but better off checking for empty `config_file`. and add fail early if `$HOME` is not available. the latter is much less likely to happen, but still..
like:
```c++
// use starts_with() in C++20
if (config_file.compare(0, 1, "~") == 0) {
    auto home = std::getenv("HOME");
    if (!home) {
        throw std::runtime_error(format("Cannot expand 'config-file': {}", config_file));
    }
    config_file.replace(0, 1, home);
}
```

i made couple changes:

1. use `compare()` instead of assuming `!config_file.empty()`. explained above already.
2.  throws if `$HOME` is not set, this is a fatal error. although it's quite like the "Cannot read config file" error, but a different one.
3. use `home` instead of `std::string(home)`, because of the 5th usage in https://en.cppreference.com/w/cpp/string/basic_string/replace

Reply:
Yup, good suggestion. Will fix shortly